File: CS/Helpers/XpoHelper.cs
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;
using System.Configuration;
using System.Reflection;

public static class XpoHelper {
    private const string ConnectionStringName = "ConnStr";

    static XpoHelper() {
        UpdateDatabase();
    }

    public static Session GetNewSession() {
        return new Session(DataLayer);
    }

    public static UnitOfWork GetNewUnitOfWork() {
        return new UnitOfWork(DataLayer);
    }

    private readonly static object lockObject = new object();

    static volatile IDataLayer fDataLayer;
    static IDataLayer DataLayer {
        get {
            if (fDataLayer == null) {
                lock (lockObject) {
                    if (fDataLayer == null) {
                        fDataLayer = GetDataLayer();
                    }
                }
            }
            return fDataLayer;
        }
    }

    private static IDataLayer GetDataLayer() {
        XpoDefault.Session = null;
        XPDictionary dict = new ReflectionDictionary();
        dict.GetDataStoreSchema(Assembly.GetExecutingAssembly());
        return new ThreadSafeDataLayer(dict, XpoDefault.GetConnectionProvider(ConfigurationManager.ConnectionStrings[ConnectionStringName].ConnectionString,
            AutoCreateOption.None));
    }

    static void UpdateDatabase() {
        using (IDataLayer dal = XpoDefault.GetDataLayer(ConfigurationManager.ConnectionStrings[ConnectionStringName].ConnectionString,
            AutoCreateOption.DatabaseAndSchema)) {
                using (Session session = new Session(dal)) {
                    Assembly asm = Assembly.GetExecutingAssembly();
                    session.UpdateSchema(asm);
                    session.CreateObjectTypeRecords(asm);
                }
        }
    }
}

File: VB/Helpers/XpoHelper.vb
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata
Imports System.Configuration
Imports System.Reflection

Public NotInheritable Class XpoHelper

	Private Sub New()
	End Sub
	Private Const ConnectionStringName As String = "ConnStr"

	Shared Sub New()
		UpdateDatabase()
	End Sub

	Public Shared Function GetNewSession() As Session
		Return New Session(DataLayer)
	End Function

	Public Shared Function GetNewUnitOfWork() As UnitOfWork
		Return New UnitOfWork(DataLayer)
	End Function

	Private ReadOnly Shared lockObject As New Object()

'INSTANT VB TODO TASK: There is no VB equivalent to 'volatile':
'ORIGINAL LINE: static volatile IDataLayer fDataLayer;
	Private Shared fDataLayer As IDataLayer
	Private Shared ReadOnly Property DataLayer() As IDataLayer
		Get
			If fDataLayer Is Nothing Then
				SyncLock lockObject
					If fDataLayer Is Nothing Then
						fDataLayer = GetDataLayer()
					End If
				End SyncLock
			End If
			Return fDataLayer
		End Get
	End Property

	Private Shared Function GetDataLayer() As IDataLayer
		XpoDefault.Session = Nothing
		Dim dict As XPDictionary = New ReflectionDictionary()
		dict.GetDataStoreSchema(System.Reflection.Assembly.GetExecutingAssembly())
		Return New ThreadSafeDataLayer(dict, XpoDefault.GetConnectionProvider(ConfigurationManager.ConnectionStrings(ConnectionStringName).ConnectionString, AutoCreateOption.None))
	End Function

	Private Shared Sub UpdateDatabase()
		Using dal As IDataLayer = XpoDefault.GetDataLayer(ConfigurationManager.ConnectionStrings(ConnectionStringName).ConnectionString, AutoCreateOption.DatabaseAndSchema)
				Using session As New Session(dal)
					Dim asm As System.Reflection.Assembly = System.Reflection.Assembly.GetExecutingAssembly()
					session.UpdateSchema(asm)
					session.CreateObjectTypeRecords(asm)
				End Using
		End Using
	End Sub
End Class
